src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+    }
+
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: bytes::BufMut,
+    {
+        let bytes = serde_json::to_vec(&self).expect("failed to encode StateUpdate");
+        buf.put(bytes.as_slice());
+    }
+
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
+        serde_json::from_slice(buf).map_err(|err| err.to_string())
+    }
+}
+
+impl mz_persist_types::columnar::Schema<StateUpdateKind> for VecU8Schema {
+    type Encoder<'a> = SimpleEncoder<'a, StateUpdateKind, Vec<u8>>;
+
+    type Decoder<'a> = SimpleDecoder<'a, StateUpdateKind, Vec<u8>>;
+
+    fn columns(&self) -> DynStructCfg {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::columns(&())
+    }
+
+    fn decoder<'a>(&self, cols: ColumnsRef<'a>) -> Result<Self::Decoder<'a>, String> {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::decoder(cols, |val, ret| {
+            *ret = StateUpdateKind::decode(val).expect("should be valid StateUpdateKind")
+        })
+    }
+
+    fn encoder<'a>(&self, cols: ColumnsMut<'a>) -> Result<Self::Encoder<'a>, String> {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::push_encoder(cols, |col, val| {
+            let mut buf = Vec::new();
+            StateUpdateKind::encode(val, &mut buf);
+            mz_persist_types::columnar::ColumnPush::<Vec<u8>>::push(col, &buf)
+        })
+    }
+}
+
+/// Handles and metadata needed to interact with persist.
+///
+/// Production users should call [`Self::_expire`] before dropping a [`PersistHandle`] so that it
+/// can expire its leases. If/when rust gets AsyncDrop, this will be done automatically.
+#[derive(Debug)]
+pub struct PersistHandle {
+    /// Write handle to persist.
+    write_handle: WriteHandle<StateUpdateKind, (), Timestamp, Diff>,
+    /// Read handle to persist.
+    read_handle: ReadHandle<StateUpdateKind, (), Timestamp, Diff>,
+}
+
+impl PersistHandle {
+    /// Create a new [`PersistHandle`] to the catalog state associated with `environment_id`.
+    pub(crate) async fn new(persist_client: PersistClient, environment_id: Uuid) -> PersistHandle {
+        // TODO(jkosh44) Using the environment ID directly is sufficient for correctness purposes.
+        // However, for observability reasons, it would be much more readble if the shard ID was
+        // constructed as `format!("s{}CATALOG", hash(environment_id))`.
+        let shard_id = ShardId::from_str(&format!("s{environment_id}")).expect("known to be valid");
+        let (write_handle, read_handle) = persist_client
+            .open(
+                shard_id,
+                Arc::new(VecU8Schema::default()),
+                Arc::new(UnitSchema),
+                Diagnostics {
+                    shard_name: "catalog".to_string(),
+                    handle_purpose: "durable catalog state".to_string(),
+                },
+            )
+            .await
+            .expect("invalid usage");
+        PersistHandle {
+            write_handle,
+            read_handle,
```

Comment:
have we thought about whether we need a critical handle in addition to the read handle? the read handle holds back the since based on time-based leases, which can occasionally fail (as we see in incident 70). a critical since hold never times out, but the flip side is if you lose track of one the since of the shard is stuck forever. in practice, we currently have one per shard that lives in the storage controller (and if two storage controllers are live concurrently, one of them learns it it got fenced out)

it's very possible that the catalog is less sensitive to this sort of thing (the storage controller has to maintain relationships between e.g. remap and source shards) and that we're fine without it. just making sure we've thought it through one way or the other :)

Reply:
I definitely have not thought this through one way or the other, I didn't know about critical handles. I will think about this and try and come up with a rationale for one over the other.

Reply:
I think we're OK without a critical handle, here's my rationale. In practice, we only ever use the read handle during bootstrapping to read the initial state out of persist. In fact, with a smart refactor of bootstrapping, we could probably remove the read handle all together. 

My worry with the critical handle is that since we don't have clean shutdown of environmentd, we have no way of cleanly releasing the critical handle during an upgrade and will always cause the shard to be stuck forever.

Reply:
Yeah, definitely better to go without if we don't need it! I think another possible rationale is you could always recover readability by advancing the upper past the since with an empty write (and then read as_of the since).

> My worry with the critical handle is that since we don't have clean shutdown of environmentd, we have no way of cleanly releasing the critical handle during an upgrade and will always cause the shard to be stuck forever.

Fwiw, this is only true if each process gets its own critical reader id. If they all share a single id (as the controller does), then it's not an issue

Reply:
>  I think another possible rationale is you could always recover readability by advancing the upper past the since with an empty write (and then read as_of the since).

I think it even works in it's current state. The read will fail causing environmentd to panic, reboot, and then try the read again. Though I'll add an action item to update to this so it's more robust.

> Fwiw, this is only true if each process gets its own critical reader id. If they all share a single id (as the controller does), then it's not an issue

Id as in shard id, or something else? After an upgrade the new catalog will use the same shard id so this should be fine actually.

Reply:
Ah, sorry, meant CriticalReaderId for the second "id", too